Hosting Dashboard: Add active subscriptions

sirbrillig opened the pull request:
<!--
Link a related Github/Linear issue to this PR. If the PR does not immediately resolve the issue,
for example, it requires a separate deployment to production, avoid
using the "fixes" keyword. Consider using "part of" instead.
-->

Part of https://linear.app/a8c/issue/CHE-229/hosting-dashboard-migrate-existing-purchases-list-to-dashboard

## Proposed Changes

This PR adds a list of active subscriptions to the Hosting Dashboard, replacing the placeholder that was there previously.

Before             |  After
:-------------------------:|:-------------------------:
<img width="1512" height="465" alt="Screenshot 2025-08-04 at 2 07 24 PM" src="https://github.com/user-attachments/assets/e12e9337-e2f9-42f5-b64b-bf1f3498f0c7" /> | <img width="1511" height="732" alt="Screenshot 2025-08-07 at 3 15 48 PM" src="https://github.com/user-attachments/assets/08cf573e-73db-4443-9a02-8b4969ba39cb" />

> [!WARNING]
> The styles, layout, and other UI here are _definitely_ not final. We are awaiting input from designers. This is just a first pass to duplicate functionality from the calypso purchases list so that refinement can continue in trunk. I've made the url `/v2/me/billing/purchases` and the title "Active Upgrades" like we use in production currently since this list is not just subscriptions in the classic sense but also includes one-time purchases.

> [!NOTE]
> Currently the view settings (eg: filtering, sorting, or pagination) are not preserved when the page is unloaded. This is an important feature but I think it's ok to implement it separately as I'm not sure the best strategy to use. The current purchases page uses the URL to encode the view data but the Sites page in the hosting dashboard does something different and I'd like to consider the options.

> [!NOTE]
> This does not include purchases made for "Earn" features (eg: Memberships, Donations, etc.) since those have a very different system. In the calypso purchases page they are shown as a separate DataView below the main DataView but that's not very good UX. We might want to show them inline with other purchases somehow or on their own page. I believe that decision can be pushed to a later PR.

> [!IMPORTANT]
> This page currently links to calypso for the single purchase view and the "add payment method" view as those pages have not yet been migrated to the hosting dashboard. Those links will be updated at a later time.

## Implementation notes

Most of this PR is copied and pasted from the existing code in the payments page, altered for the rules of the hosting dashboard (eg: using `__()` instead of `translate()`). To review this code, you may want to compare each component against its original implementation. For example: 

- `PurchaseProduct` is a duplicate of `PurchaseItemProduct`
- `PurchaseStatus` is a duplicate of `PurchaseItemStatus`
- `PurchasePaymentMethod` is a duplicate of `PurchaseItemPaymentMethod`

Calypso's purchases page uses some functions which have very deep dependency chains, often ending up a the massive hard-coded `PLANS_LIST` array in the `@automattic/calypso-products` package. In this PR I explored each of the function calls from the original components and tried to replace them with a simpler version using the properties of the purchases themselves (eg: using server-side properties like those added in 188916-ghe-Automattic/wpcom and 190054-ghe-Automattic/wpcom to replace complex functions like `isPlan`, `isJetpackAISlug`, or `isJetpackPlan()`).

Another major difference is that the original components used the `Purchases.Purchase` object type which is constructed by the [createPurchaseObject()](https://github.com/Automattic/wp-calypso/blob/d4fc100b9c4a65a4dcf8b08dede9d0c032640e76/packages/data-stores/src/purchases/lib/assembler.ts#L4) function using the data returned by the purchases endpoint. However, that's not a common pattern in calypso; most code uses raw data from the endpoint instead which greatly simplifies adding new properties. In addition, the translation layer causes challenges with some properties by making them nested instead of flat which sometimes causes fatal errors when the "parent" properties are not set. It always means there are two versions of each property to keep in your head any time we work between the server and the client, slowing down debugging and new features. I decided to take the opportunity to _not_ duplicate `createPurchaseObject()` and instead use the raw data directly from the server.

## Testing Instructions

View http://calypso.localhost:3000/v2/me/billing/purchases and compare its UI/UX to http://calypso.localhost:3000/me/purchases 

Verify that the same purchases are shown with the same data and that the various links operate as expected (except for the caveats listed above).

You can do the same with the sites-level purchases list. Compare http://calypso.localhost:3000/v2/me/billing/purchases/example.com (using your site's URL at the end) to http://calypso.localhost:3000/purchases/subscriptions/example.com

sirbrillig committed:
Add basic fetching of user-level purchases to dashboard

sirbrillig committed:
Add DataViews with basic info

sirbrillig committed:
Remove unnecessary chrome and placeholder text

sirbrillig committed:
Adjust table for page width

sirbrillig committed:
Add additional product types to ActiveSubscription type

These will require a server-side change!

sirbrillig committed:
Add ActiveSubscriptionDescription

sirbrillig committed:
Supply sites via getSites query

sirbrillig committed:
Add site icon

sirbrillig committed:
Add expiry column

sirbrillig committed:
Add site-level route

sirbrillig committed:
Add payment method to subscription list

sirbrillig committed:
Make site-level view its own component

sirbrillig committed:
Implement isCloseToExpiration

sirbrillig committed:
Implement creditCardExpiresBeforeSubscription

sirbrillig committed:
Opt-out of filtering for non-filterable columns

sirbrillig committed:
Implement creditCardHasAlreadyExpired

sirbrillig committed:
Fix help link in ActiveSubscriptionExpiry

sirbrillig committed:
Fix locale formatting in ActiveSubscriptionExpiry

sirbrillig committed:
Add actions to DataViews

sirbrillig committed:
Prevent 'undefined' from being part of field getValue

sirbrillig committed:
Fix date comparisons (facepalm)

sirbrillig committed:
Use is_non_free_plan to look for plan description

sirbrillig committed:
Add initial work on adding Jetpack display names

sirbrillig committed:
Remove unnecessary Jetpack name functions

The backend already provides these names with exactly, or near-exactly
the same text. All this logic is redundant.

sirbrillig committed:
Add PaymentMethodImage

sirbrillig committed:
Add backup payment method notice

sirbrillig committed:
Replace is_non_free_plan with is_plan

sirbrillig committed:
Fix Popover for BackupPaymentMethodNotice

sirbrillig committed:
Move payment method fetching to data layer

sirbrillig committed:
Fix call to getFields in site-level view

sirbrillig committed:
Move getAddPaymentMethodUrlFor up to data-view-shared

This will co-locate it with the other get...Url methods

sirbrillig committed:
Add OwnerInfo to purchase title

sirbrillig committed:
Handle transferred ownership subscriptions

sirbrillig committed:
Add isA4ATemporarySitePurchase helper

sirbrillig committed:
Add queryOptions to new queries

sirbrillig committed:
Move bill periods and akisment plans to enum constants

sirbrillig committed:
Add automattic/urls as an allowed package

See discussion here: https://a8c.slack.com/archives/C0982082MSR/p1753982850724449

sirbrillig committed:
Use JETPACK_CONTACT_SUPPORT in OwnerInfo

sirbrillig committed:
Remove className from tags

sirbrillig committed:
Switch from useSuspenseQuery to useQuery

This way we get a loading state and can pass it to DataViews

Also this fixes a bug in the site-level view by adding transferredPurchases.

sirbrillig committed:
Add missing constants file

sirbrillig committed:
Wrap ActiveSubscriptionDescription in div

sirbrillig committed:
Use specific site icons for jetpack, akismet, no site, etc.

fushar commented:
Thanks for the PR; @Automattic/lego is planning to review this PR as a team tomorrow APAC morning 😄

chrismccluskey reviewed (commented):
I tested this code in Safari, Chrome and Firefox on OS X 15.5 (Intel) without any issues. The code looks clean and devoid of errors to me and my IDE but I'll defer to Lego for their approval.

taipeicoder commented on client/dashboard/app/router.tsx:
```diff
@@ -569,6 +569,17 @@ const activeSubscriptionsRoute = createRoute( {
 	)
 );
 
+const activeSubscriptionsSiteRoute = createRoute( {
```
Team review: Consider adding `loader` to the route to fetch data.

sirbrillig replied:
Yes, good idea. I just need to grok the syntax needed to do that... 👀

taipeicoder commented on client/dashboard/app/queries/me-transferred-purchases.ts:
```diff
@@ -0,0 +1,12 @@
+import { queryOptions } from '@tanstack/react-query';
+import { fetchTransferredPurchases } from '../../data/me-transferred-purchases';
+
+const queryKey = 'transferred-purchases';
```
Team review: Consider inlining constant strings for query keys for better type inference.

taipeicoder commented on client/dashboard/app/queries/me-active-subscriptions.ts:
```diff
@@ -0,0 +1,8 @@
+import { queryOptions } from '@tanstack/react-query';
+import { fetchActiveSubscriptionsForUser } from '../../data/me-active-subscriptions';
+
+export const activeSubscriptionsQuery = ( { siteId }: { siteId?: string | number } ) =>
```
Team review: siteId should have the type `number`. By ensuring it's a number, we reduce missed caching. For instance, queryKey [ `site`, 123 ] will not match with [ `site`, `a.example.com` ].

In this case, using siteSlug might make more sense so that users can see their site slug in the URL instead of the site ID.

sirbrillig replied:
You're right. But this is not trivial to do.

ID is the most reliable since it will never be affected by custom domains (you never know what sort of "site slug" you're going to have; is it example.wordpress.com, example.wpstaging.com, or example.com?). 

However, sometimes all we have is the site slug from the URL and it would take another fetch to another endpoint to turn it into an ID.

🤔 I'm leaving these as `string | number` for now but I left a note to figure out a better option later.

taipeicoder replied:
In that case could we rename it to siteIdOrSlug?

arthur791004 replied:
I think if the slug might not always work, we should treat it explicitly as an ID and avoid assuming it could be either an ID or a string.

So, I’d suggest making it either `siteId: number` or `siteSlug: string`

fushar replied:
I agree to keep it as `siteId: number` for now to avoid leaking complexity to existing code... and we can follow-up later if we want to show site slug in the URL. 🙂

taipeicoder commented on client/dashboard/data/me-active-subscriptions.ts:
```diff
@@ -0,0 +1,188 @@
+	maximum_units?: undefined | null | number;
+	minimum_price: number;
+	minimum_price_display: string;
+	minimum_price_monthly_display?: string | null | undefined;
```
Team review: `undefined` is redundant for optional properties.

taipeicoder commented on client/dashboard/data/me-active-subscriptions.ts:
```diff
@@ -0,0 +1,188 @@
+} ): Promise< ActiveSubscription[] > {
+	if ( options?.siteId ) {
+		return await wpcom.req.get( {
+			path: `/sites/${ encodeURIComponent( options.siteId ) }/purchases`,
```
Team review: [nit] Since the fetch function uses different functions, consider create separate fetch functions.

taipeicoder commented on client/dashboard/data/me-payment-methods.ts:
```diff
@@ -0,0 +1,102 @@
+export const fetchPaymentMethods = (
+	type: PaymentMethodRequestType,
+	expired: boolean
+): StoredPaymentMethod[] =>
```
Team review: fetchPaymentMethods should return a Promise. This has the advantage for `select` to benefit from type inference.

taipeicoder commented on client/dashboard/app/queries/me-payment-methods.ts:
```diff
@@ -0,0 +1,50 @@
+		queryKey: [ storedPaymentMethodsQueryKey, type, expired ],
+		queryFn: () => fetchPaymentMethods( type, expired ),
+		enabled: ! isLoggedOut,
+		select: ( data: undefined | Array< StoredPaymentMethod > ) =>
```
Team review: Consider removing type definitions here in favor of add them in `fetchPaymentMethods`. That way the type is inferred into the select function.

taipeicoder commented on client/dashboard/data/me-transferred-purchases.ts:
```diff
@@ -0,0 +1,9 @@
+import wpcom from 'calypso/lib/wp';
+import type { ActiveSubscription } from './me-active-subscriptions';
+
+export function fetchTransferredPurchases(): ActiveSubscription[] {
```
Team review: Should return a Promise here as well.

taipeicoder commented on client/dashboard/me/active-subscriptions/active-subscription-description.tsx:
```diff
@@ -0,0 +1,225 @@
+	getUrlForSiteLevelView,
+}: {
+	purchase: ActiveSubscription;
+	site?: { name: string; slug: string; ID: number };
```
Team review: Consider using the Site type we have for the HD.

taipeicoder commented on client/dashboard/me/active-subscriptions/active-subscription-description.tsx:
```diff
@@ -0,0 +1,225 @@
+								/>
+							),
+							link: (
+								<a
```
Team review: [nit] Consider using ExternalLink for consistency on external links handling.

sirbrillig replied:
This doesn't work when combined with `createInterpolateElement()` in TypeScript. It can't tell that the `children` prop will be added later.

```
Property 'children' is missing in type '{ href: string; rel: string; title: string; }' but required in type 'Omit<WordPressComponentProps<ExternalLinkProps, "a", false>, "target">'.
```

arthur791004 replied:
We can pass `children={ null }` or add `// @ts-expect-error children prop is injected by createInterpolateElement` to bypass the error. I think it’s more important to keep the external link consistent, even if it causes this kind of type error.

taipeicoder commented on client/dashboard/me/active-subscriptions/active-subscription-description.tsx:
```diff
@@ -0,0 +1,225 @@
+		);
+	}
+
+	return <div>{ productType }</div>;
```
Team review: [question] Why does it need to be wrapped within a `<div>`? Also, for texts, consider using the `<Text>` component.

taipeicoder commented on client/dashboard/me/active-subscriptions/active-subscription-expiry.tsx:
```diff
@@ -0,0 +1,323 @@
+
+export function ActiveSubscriptionExpiry( {
+	purchase,
+	isJetpack,
```
Team review: [suggestion] Should we avoid using props such as isJetpack and isDisconnectedSite in favor of passing site as a prop and determined from using utility functions that are shared across HD?

taipeicoder commented on client/dashboard/me/active-subscriptions/active-subscription-expiry.tsx:
```diff
@@ -0,0 +1,323 @@
+			sprintf(
+				// translators: date is a formatted date, amount is a currency amount, and excludeTaxStringAbbreviation is something like "excludes VAT"
+				__(
+					'Free trial ends on <span>%(date)s</span>, renews automatically at %(amount)s <abbr>%(excludeTaxStringAbbreviation)s</abbr>'
```
Team review: Consider moving HTML tags inside components instead of the string. For instance: instead of `<abbr>%(excludeTaxStringAbbreviation)s</abbr>`, use `<excludeTaxStringAbbreviation />`.

arthur791004 replied:
See https://github.com/Automattic/wp-calypso/pull/103700#discussion_r2108720909.

taipeicoder commented on client/dashboard/me/active-subscriptions/active-subscription-payment-method.tsx:
```diff
@@ -0,0 +1,78 @@
+export function ActiveSubscriptionPaymentMethod( {
+	purchase,
+	isDisconnectedSite,
+	getAddPaymentMethodUrlFor,
```
Team review: [nit] Consider moving getAddPaymentMethodUrlFor out instead of passing it as a prop.

sirbrillig replied:
I figured this would be helpful for now so that the URLs we are going to replace as part of this migration are all in one place.

taipeicoder commented on client/dashboard/me/active-subscriptions/constants.ts:
```diff
@@ -0,0 +1,53 @@
+export const SubscriptionBillPeriod = {
```
Team review: [nit] Consider naming constants in all caps. SubscriptionBillPeriod -> SUBSCRIPTION_BILL_PERIOD.

fushar replied:
Wait actually the existing constants (enums) are already in SentenceCaps (https://github.com/Automattic/wp-calypso/blob/d841bebb5039600539255210541cf3cc6aff9b70/client/dashboard/data/constants.ts), so I think this is OK.

sirbrillig replied:
I'm not sure. The existing constants like `DotcomFeatures` or `DotcomPlans` don't follow that pattern. I think I personally like camel case for the group.

arthur791004 replied:
Ah yes, I forgot that. However, it does seem a bit inconsistent to me that the outer name uses SentenceCaps, while the properties do not 🙈

taipeicoder commented on client/dashboard/me/active-subscriptions/data-view-shared.tsx:
```diff
@@ -0,0 +1,543 @@
+import { formatNumber } from '@automattic/number-formatters';
+import { JETPACK_CONTACT_SUPPORT } from '@automattic/urls';
+import { Popover, Icon } from '@wordpress/components';
+import { type SortDirection, type View, type Fields } from '@wordpress/dataviews';
```
Team review: [nit]
```suggestion
import type {  SortDirection, View, Fields } from '@wordpress/dataviews';
```

taipeicoder commented on client/dashboard/me/active-subscriptions/data-view-shared.tsx:
Team review: [non-blocking] This file seems to lack focus as it's a combination of utility function and also components.

Team review: Consider renaming this file to dataviews.tsx for consistency. We should do this going forward for all out dataviews logic. Or split to dataviews/{ views, fields } if too complex.

taipeicoder commented on client/dashboard/me/active-subscriptions/util.ts:
Team review: [nit] Consider renaming the file to /utils.

fushar replied:
Also consider moving them to `client/dashboard/utils/` for consistency with existing pattern 😄

taipeicoder commented on client/dashboard/me/active-subscriptions/active-subscription-expiry.tsx:
```diff
@@ -0,0 +1,323 @@
+	isTemporarySitePurchase,
+	isA4ATemporarySitePurchase,
+	isRecentMonthlyPurchase,
+	getRelativeTimeString,
```
Team review: [nit] Consider using the HD component `<TimeSince>` instead of `getRelativeTimeString()`.

sirbrillig replied:
`TimeSince` only shows custom text for dates in the past and only for dates within the past 7 days. Very few subscription expiration dates fall into that category. Most will read "in 2 years" or "in 5 months" or "3 months ago".

arthur791004 replied:
I'm curious is it possible to use `date-fns` to handle this?

For example
```js
import { formatDistanceToNow, add } from 'date-fns';

// Future date: "in 2 years"
const in2Years = formatDistanceToNow(add(new Date(), { years: 2 }), { addSuffix: true, locale });
// → "in about 2 years"

// Future date: "in 5 months"
const in5Months = formatDistanceToNow(add(new Date(), { months: 5 }), { addSuffix: true, locale });
// → "in 5 months"

// Past date: "3 months ago"
const threeMonthsAgo = formatDistanceToNow(sub(new Date(), { months: 3 }), { addSuffix: true, locale });
// → "3 months ago"

// Past date: "3 days ago"
const threeDaysAgo = formatDistanceToNow(sub(new Date(), { days: 3 }), { addSuffix: true, locale });
// → "3 days ago"
```

But this is not the scope of this PR, and we can revisit whether it works in the future.

sirbrillig replied:
Yes, I did this work before discovering we could use `date-fns` in the hosting dashboard. I'd be happy to switch.

taipeicoder reviewed (commented):
Thanks for the PR @sirbrillig. @Automattic/lego took the opportunity to review this as a team, with the goal of improving the Hosting Dashboard README in the process. I left inline reviews where it made sense, and there are some general notes that I'll jot down below.

* We are already using the term "purchase" in other parts of the dashboard logic. Consider continuing using that term and only use “Active Subscriptions” in the UI.
* With that in mind, let’s restructure the data and query layers as follows:
  * client/dashboard/data:
    * purchase.ts - Extract the Purchase type from site-purchases.ts
    * site-purchases.ts - For /sites/${ siteId }/purchases (actually this already exists)
    * me-purchases.ts - For /me/purchases and /me/purchases/transferred
    * me-payment-methods.ts - For /me/payment-methods
  * client/dashboard/app/queries:
    * site-purchases.ts - For sitePurchasesQuery() (file already exists)
    * me-purchases.ts - For userPurchasesQuery() and userTransferredPurchasesQuery()
    * me-payment-methods.ts - For userPaymentMethodsQuery()
* Remove all `isLoggedOut` logic since it doesn't seem to be used.

sirbrillig committed:
Rename Active Subscriptions with Purchases

sirbrillig committed:
Refactor queries to split site and user

sirbrillig committed:
Clarify properties of Purchase type

sirbrillig committed:
Update transferredPurchases to remove isLoggedOut

sirbrillig committed:
Clarify types of site purchase queries

sirbrillig committed:
Clean up redundant undefined types in Purchase

sirbrillig committed:
Fix return type of fetchPaymentMethods

sirbrillig committed:
Remove unnecessary parts of paymentMethodsQuery

sirbrillig committed:
Rename data-view-shared to dataviews

sirbrillig committed:
Move getRelativeTimeString/isWithinLast/isWithinNext to utils/datetime

sirbrillig committed:
Move getDateFromCreditCardExpiry to utils/datetime

sirbrillig committed:
Move constants and utils to higher level directories

sirbrillig committed:
Clean up PurchaseProduct a bit

sirbrillig committed:
Fix links to purchases URLs

sirbrillig committed:
Add is_jetpack_plan_or_product to Purchase type

sirbrillig committed:
Handle isJetpack inline in PurchaseStatus

arthur791004 commented on client/dashboard/app/queries/me-transferred-purchases.ts:
```diff
@@ -0,0 +1,9 @@
+
+export function transferredPurchasesQuery() {
+	return queryOptions( {
+		queryKey: [ 'transferred-purchases' ],
```
It would be better to add `me` to the query key so we can invalidate the all queries when the user changes.

```
queryKey: [ 'me', 'transferred-purchases' ],
```

arthur791004 commented on client/dashboard/app/queries/site-purchases.ts:
```diff
@@ -1,7 +1,14 @@
+ * reliable since it will never be affected by custom domains. However,
+ * sometimes all we have is the site slug from the URL.
+ */
+
```
[nit]: extra line?

arthur791004 commented on client/dashboard/data/purchase.ts:
```diff
@@ -0,0 +1,232 @@
+
+export interface RawPurchasePriceTierEntry extends PriceTierEntry {
+	minimum_price_monthly_display: never;
+	maximum_price_monthly_display: never;
```
I'm curious why it's `never` 🤔

sirbrillig replied:
Weird! I just copied these types from calypso. I think this was never filled out. Fixing.

arthur791004 commented on client/dashboard/data/purchase.ts:
```diff
@@ -0,0 +1,232 @@
+		| 'paypal'
+		| 'emergent-paywall'
+		| 'brazil-tef'
+		| string;
```
Adding a string here feels a bit off, since it means the payment type could be any arbitrary string 🤔

arthur791004 commented on client/dashboard/me/purchases/dataviews.tsx:
```diff
@@ -0,0 +1,544 @@
+	layout: {},
+};
+
+function getDisplayName( item: Purchase ): string {
```
nit: Looks like we can call it `purchase` for consistency

arthur791004 commented on client/dashboard/me/purchases/site.tsx:
```diff
@@ -0,0 +1,84 @@
+			}
+		>
+			<div>
+				<a href="/v2/me/billing/purchases">{ __( 'View all active subscriptions' ) }</a>
```
We should use `Link` from `@tanstack/react-router`

arthur791004 commented on client/dashboard/me/purchases/purchase-status.tsx:
```diff
@@ -0,0 +1,325 @@
+			sprintf(
+				// translators: date is a formatted date, amount is a currency amount, and excludeTaxStringAbbreviation is something like "excludes VAT"
+				__(
+					'Free trial ends on <span>%(date)s</span>, renews automatically at %(amount)s <abbr>%(excludeTaxStringAbbreviation)s</abbr>'
```
There's no good reason to both have a markup and placeholder. See https://github.com/Automattic/wp-calypso/pull/103700#discussion_r2108720909.

fushar commented on client/dashboard/app/queries/me-payment-methods.ts:
```diff
@@ -0,0 +1,40 @@
+
+const storedPaymentMethodsQueryKey = 'payment-methods';
+
+export const paymentMethodsQuery = ( {
```
I know it might sound annoying, but for the sake of consistency, can we prefix all queries under `me-*` with `user`, and use `['me', ... ` query key array prefix with correct hierarchy? So:

- `userPaymentMethodsQuery()` -> `['me', 'payment-methods']`
- `userPurchasesQuery()` -> `['me', 'purchases']` (already correct, thanks!)
- `userTransferredPurchasesQuery()` -> `['me', 'purchases', 'transferred']`

fushar commented on client/dashboard/app/queries/me-payment-methods.ts:
```diff
@@ -0,0 +1,40 @@
+import { fetchPaymentMethods } from '../../data/me-payment-methods';
+import type { PaymentMethodRequestType } from '../../data/me-payment-methods';
+
+const storedPaymentMethodsQueryKey = 'payment-methods';
```
[nit] We think we can just inline the key in the query options instead of extracting it in a variable.

fushar commented on client/dashboard/app/queries/me-transferred-purchases.ts:
```diff
@@ -0,0 +1,9 @@
+import { queryOptions } from '@tanstack/react-query';
+import { fetchTransferredPurchases } from '../../data/me-transferred-purchases';
+
+export function transferredPurchasesQuery() {
```
[nit] I think we can just put it in the same `me-purchases.ts` file to group related queries.

fushar commented on client/dashboard/data/me-transferred-purchases.ts:
```diff
@@ -0,0 +1,9 @@
+import wpcom from 'calypso/lib/wp';
+import type { Purchase } from './purchase';
+
+export async function fetchTransferredPurchases(): Promise< Purchase[] > {
```
[nit] I think we can just put this in the same `me-purchases.ts` file to group related endpoints.

fushar reviewed (commented):
This is looking good and heading to the right direction 😄 left some more consistency-related comments. Thanks for working on this!

fushar reviewed (commented):
Consistency: can we rename `me/purchases/` to `me/billing-purchases/`?

We started to use this pattern, where the routes under the same tab should share the same prefix. Existing `billing-history` is already correct, but `payment-methods` and `tax-details` are not. (I guess we just didn't have time to clean up these routes)

sirbrillig committed:
Put transferredPurchasesQuery with userPurchasesQuery

sirbrillig committed:
Rename paymentMethodsQuery to userPaymentMethodsQuery

sirbrillig committed:
Fix some types in Purchase

sirbrillig committed:
Rename item to purchase in dataviews

sirbrillig committed:
Use Link for anchor tag in PurchasesForSite

sirbrillig committed:
Reword "Active Subscriptions" to "Active Upgrades"

This better matches the existing purchases page and is more clear that
not all things listed are conceptually "subscriptions". For example,
while one-time purchases are subscriptions under the hood, they have
none of the things that most people would associate with a subscription:
renewals, expiry dates, etc.

sirbrillig committed:
Rename purchases directory to billing-purchases

sirbrillig committed:
Add 'upi' to possible payment_type

sirbrillig committed:
Use ExternalLink in PurchaseProduct

sirbrillig committed:
Standardize on using number for sitePurchasesQuery

sirbrillig committed:
Do not use component and string interpolation together

sirbrillig committed:
Change some missed "active subscriptions" text to "active upgrades"

sirbrillig committed:
Fix impossible type error

sirbrillig committed:
Add loader for purchasesRoute and purchasesSiteRoute

sirbrillig committed:
Make getUrlForSiteLevelView use site slug instead of ID

sirbrillig committed:
Use siteBySlugQuery in PurchasesForSite

sirbrillig commented:
@fushar @taipeicoder @arthur791004 Thank you for all the feedback! It was really helpful to figure out how this should all be architected. I believe I've addressed all the comments you've provided. Please let me know if you have additional thoughts.

taipeicoder reviewed (approved):
Looks good on my end 👍

fushar reviewed (approved):
Tested the feature, and re-reviewed the code. It's looking really great now. Thanks for the hard work here 🙏 Let's ship this!

a8ci18n commented:
This Pull Request is now available for translation here: https://translate.wordpress.com/deliverables/19751636

Some locales (Hebrew, Japanese) have been temporarily machine-translated due to translator availability. All other translations are usually ready within a few days. Untranslated and machine-translated strings will be sent for translation next Monday and are expected to be completed by the following Friday.

Thank you @sirbrillig for including a screenshot in the description! This is really helpful for our translators.
